src/Bech32.mo

```diff
@@ -0,0 +1,296 @@
+        };
+        case _ {
+          #err ("Failed to decode HRP.");
+        };
+    };
+  };
+
+  // Expand HRP for checksum computations by grouping together the first 3 bits
+  // of all characters, then a zero then the last 5 bits of all characters.
+  // [a, b] => [a[:3], b[:3]] + [0] + [a[3:], b[3:]].
+  func expandHrp(hrp : [Nat8]) : [Nat8] {
+    let hrpSize = hrp.size();
+    let outputSize = hrpSize * 2 + 1;
+    let output = Array.init<Nat8>(outputSize, 0);
+
+    var i : Nat = 0;
+    for (currHrp in hrp.vals()) {
+      output[i] := currHrp >> 5;
+      output[i + hrpSize + 1] := currHrp & 0x1f;
+      i += 1;
+    };
+
+    return Array.freeze(output);
+  };
+
+  // Constant value associated to the given encoding.
+  func encodingConstant(encoding : Encoding) : Nat32 {
+    return switch(encoding) {
+      case (#BECH32) {
+        1
+      };
+      case (#BECH32M) {
+        0x2bc830a3
+      };
+    };
+  };
+
+  // Compute the checksum values for given hrp and data.
+  func createChecksum(hrp : [Nat8], data : [Nat8], encoding : Encoding) : [Nat8] {
+    let expandedHrp : [Nat8] = expandHrp(hrp);
+
+    // Merge expandedHrp and data arrays and append 6 zeroes to get
+    // [expandedHrp..., data..., 0, 0, 0, 0, 0, 0].
+    let polyModValues : Buffer.Buffer<Nat8> = Buffer.Buffer<Nat8>(
+      expandedHrp.size() + data.size() + 6);
+
+    for (val in expandedHrp.vals()) {
+      polyModValues.add(val);
+    };
+    for (val in data.vals()) {
+      polyModValues.add(val);
+    };
+
+    // Pad with 6 zeros.
+    for (_ in Iter.range(0, 5)) {
+      polyModValues.add(0);
+    };
+
+    let mod : Nat32 = polymod(polyModValues.toArray())
+      ^ encodingConstant(encoding);
+
+    // Convert the 5-bit groups in mod to checksum data.
+    return Array.tabulate<Nat8>(6, func (i) {
+      Nat8.fromIntWrap(Nat32.toNat(
+        (mod >> (5 * (5 - Nat32.fromIntWrap(i)))) & 31))
+    });
+  };
+
+  func verifyChecksum(hrp : [Nat8], values : [Nat8])
+    : Result.Result<Encoding, Text> {
+
+    let expandedHrp : [Nat8] = expandHrp(hrp);
+
+    // Merge expandedHrp and values arrays.
+    let polyModValues : Buffer.Buffer<Nat8> = Buffer.Buffer<Nat8>(
+      expandedHrp.size() + values.size());
+
+    for (val in expandedHrp.vals()) {
+      polyModValues.add(val);
+    };
+    for (val in values.vals()) {
+      polyModValues.add(val);
+    };
+
+    let check : Nat32 = polymod(polyModValues.toArray());
+
+    return if (check == encodingConstant(#BECH32)) {
+      #ok (#BECH32)
+    } else if (check == encodingConstant(#BECH32M)) {
+      #ok (#BECH32M)
+    } else {
+      #err("Checksum verification failed.")
+    };
+  };
+
+  // Compute 6 5-bit values that make the checksum zero. Input values are
+  // coefficients of a polynomial over GF(32) with an implicit 1 in front.
+  // Returns the values packed inside a Nat32.
+  func polymod(values : [Nat8]) : Nat32 {
+    var c : Nat32 = 1;
+
+    for (_value in values.vals()) {
```

Comment:
I think preceding a variable with an underscore doesn't indicate whether it's used. If it weren't used it would just be `for (_ in values.vals())`. In any case,  the `_` prefix  was due to an unnecessary intermediate assignment and I just got rid of it.

Reply:
Yeah, you are right. A leading underscore is sometimes used to identify private variables but that wouldn't be useful here.